crates/api/src/transport.rs

```diff
@@ -0,0 +1,350 @@
+    /// On success, this function returns bytes that should be
+    /// sent as a preflight message for additional connection validation.
+    /// (The preflight data should be sent even if it is zero length).
+    pub fn peer_connect(&self, peer: Url) -> K2Result<bytes::Bytes> {
+        for h in self.mod_map.lock().unwrap().values() {
+            h.peer_connect(peer.clone())?;
+        }
+        for h in self.space_map.lock().unwrap().values() {
+            h.peer_connect(peer.clone())?;
+        }
+        self.handler.peer_connect(peer.clone())?;
+        let preflight = self.handler.preflight_gather_outgoing(peer)?;
+        let enc = (K2Proto {
+            ty: k2_proto::Ty::Preflight as i32,
+            data: preflight,
+            space: None,
+            module: None,
+        })
+        .encode()?;
+        Ok(enc)
+    }
+
+    /// Call this whenever a connection is closed.
+    pub fn peer_disconnect(&self, peer: Url, reason: Option<String>) {
+        for h in self.mod_map.lock().unwrap().values() {
+            h.peer_disconnect(peer.clone(), reason.clone());
+        }
+        for h in self.space_map.lock().unwrap().values() {
+            h.peer_disconnect(peer.clone(), reason.clone());
+        }
+        self.handler.peer_disconnect(peer, reason);
+    }
+
+    /// Call this whenever data is received on an open connection.
+    pub fn recv_data(&self, peer: Url, data: bytes::Bytes) -> K2Result<()> {
+        let data = K2Proto::decode(&data)?;
+        let ty = data.ty();
+        let K2Proto {
+            space,
+            module,
+            data,
+            ..
+        } = data;
+
+        match ty {
+            k2_proto::Ty::Unspecified => Ok(()),
+            k2_proto::Ty::Preflight => {
+                self.handler.preflight_validate_incoming(peer, data)
+            }
+            k2_proto::Ty::Notify => {
+                if let Some(space) = space {
+                    let space = SpaceId::from(space);
+                    if let Some(h) = self.space_map.lock().unwrap().get(&space)
+                    {
+                        h.recv_space_notify(peer, space, data);
+                    }
+                }
+                Ok(())
+            }
+            k2_proto::Ty::Module => {
+                if let (Some(space), Some(module)) = (space, module) {
+                    let space = SpaceId::from(space);
+                    if let Some(h) = self
+                        .mod_map
+                        .lock()
+                        .unwrap()
+                        .get(&(space.clone(), module.clone()))
+                    {
+                        h.recv_module(peer, space, module, data);
+                    }
+                }
+                Ok(())
+            }
+            k2_proto::Ty::Disconnect => {
+                let reason = String::from_utf8_lossy(&data).to_string();
+                Err(K2Error::other(format!("Remote Disconnect: {reason}")))
```

Comment:
The rest of this code *can* error in the higher level handlers and that always indicates the connection should be closed too. I'm totally up for having a more complex return type, but the functionality would then boil down to "close the connection if you get a Close variant OR if you get an error" so I worry it would just be adding complexity for no reason.

Reply:
Just because the remote sent us a graceful disconnect message doesn't mean they actually disconnected from us : ) We need to actually disconnect from them so at least *we* are in a consistent state.